Refactor Test&Troubleshoot (#2666)
astronomer/docs: astro/cli/troubleshoot-your-local-airflow-environment.md

---
sidebar_label: 'Troubleshoot your Airflow environment'
title: 'Troubleshoot your local Airflow environment using the Astro CLI'
id: troubleshoot-your-local-airflow-environment
description: Run commands in your local Airflow environment to troubleshoot running DAGs and tasks.
---

> Some issues with a local Astro project can be solved only after running your local Airflow environment with `astro dev start`. For example, one of your DAGs might cause an error in your Airflow scheduler, which you can only know by triggering the task and checking your scheduler logs.

Author: ```suggestion
Some issues with a local Astro project can be solved only after running your local Airflow environment with `astro dev start`.
```

)
```

## Troubleshoot KubernetesPodOperator issues

View local Kubernetes logs to troubleshoot issues with Pods that are created by the KubernetesPodOPerator. See [Test and Troubleshoot the KubernetesPodOperator Locally](https://docs.astronomer.io/learn/kubepod-operator#run-the-kubernetespodoperator-locally).

## Hard reset your local environment

In most cases, [restarting your local project](develop-project.md#restart-your-local-environment) is sufficient for testing and making changes to your project. However, it is sometimes necessary to kill your Docker containers and metadata database for testing purposes. To do so, run the following command:

```sh
astro dev kill
```

This command forces your running containers to stop and deletes all data associated with your local Postgres metadata database, including Airflow connections, logs, and task history.

## Troubleshoot dependency errors

When dependency errors occur, the error message that is returned often doesn't contain enough information to help you resolve the error. To retrieve additional error information, you can review individual operating system or python package dependencies inside your local Docker containers.

> For example, if your `packages.txt` file contains the `openjdk-8-jdk`, `gcc`, `g++`, or `libsas12-dev` packages and you receive build errors after running `astro dev start`, you can enter the container and install the packages manually to review additional information about the errors.

Author: ```suggestion
For example, if your `packages.txt` file contains several packages and you receive build errors after running `astro dev start`, you can enter the container and install the packages manually to review additional information about the errors.
```

> 

Author: I actually think the package names are a bit distracting, so I went for a more generic example.

> 1. Open the `requirements.txt` and `packages.txt` files for your project and remove the references to the packages that are returning error messages.

Reviewer: Since we are talking about `packages.txt` in the example, so we don't want to mention `requirements.txt`
```suggestion
1. Open the `packages.txt` files for your project and remove the references to the packages that are returning error messages.
```

2. Run the following command to build your Astro project into a Docker image and start a local Docker container for each Airflow component:

    ```sh
    astro dev start
    ```

3. Run the following command to open a bash terminal in your scheduler container:

    ```sh
    astro dev bash --scheduler
    ```

4. In the bash terminal for your container, run the following command to install a package and review any error messages that are returned:

    ```bash
    apt-get install <package-name>
    ```
    For example, to install the GNU Compiler Collection (GCC) compiler, you would run:

    ```bash
    apt-get install gcc
    ```

> 5. Open the `requirements.txt` and `packages.txt` files for your project and add the package references you removed in step 1 one by one until you find the package that is the source of the error.

Reviewer: ```suggestion
5. Open the `packages.txt` files for your project and add the package references you removed in step 1 one by one until you find the package that is the source of the error.
```

## Override the Astro CLI Docker Compose file

The Astro CLI uses a default set of [Docker Compose](https://docs.docker.com/compose/) configurations to define and run local Airflow components. For advanced testing cases, you might need to override these default configurations. For example, you might need to:

- Add extra containers to mimic services that your Airflow environment needs to interact with locally, such as an SFTP server.
- Change the volumes mounted to any of your local containers.

:::info

The Astro CLI does not support overrides to environment variables that are required globally. For the list of environment variables that Astro enforces, see [Global environment variables](platform-variables.md). To learn more about environment variables, read [Environment variables](environment-variables.md).

:::

1. Reference the Astro CLI's default [Docker Compose file](https://github.com/astronomer/astro-cli/blob/main/airflow/include/composeyml.yml) (`composeyml.yml`) and determine one or more configurations to override.

> 2. Add a `docker-compose.override.yml` file to your Astro project.

Reviewer: ```suggestion
2. Add a `docker-compose.override.yml` file to your Astro project at the same level as your Dockerfile.
```

3. Specify your new configuration values in `docker-compose.override.yml` file using the same format as in `composeyml.yml`.

For example, to add another volume mount for a directory named `custom_dependencies`, add the following to your `docker-compose.override.yml` file:

```yaml
version: "3.1"
services:
  scheduler:
    volumes:
      - /home/astronomer_project/custom_dependencies:/usr/local/airflow/custom_dependencies:ro
```

Make sure to specify `version: "3.1"` and follow the format of the source code file linked above.

> To see your override file live in your local Airflow environment, run the following command to see the file in your scheduler container:

Reviewer: ```suggestion
Run the following command to see the override file in your scheduler container:
```
